amd-ras: Modify to call apml alert APIs to monitor RAS alerts

akky16 opened the pull request:
-Update apml_manager to call APML APIs to register for uevent from apml alertl driver.
- APML library  provides API to register, monitor and unregister for the alerts on RAS and temperature from apml alertl driver.
- The alertEventHandler can handle alerts from socket 0 and 1. 
- On alerts, the monitor_ras_alert returns socket & die-id and alertl source information.

abinayaddhandapani commented on src/apml_manager.cpp:
```diff
@@ -1410,71 +1334,15 @@ bool Manager::decodeInterrupt(uint8_t socNum)
-    if (read_sbrmi_ras_status(socNum, &buf) == OOB_SUCCESS)
-    {
-        lg2::debug("Read RAS status register. Value: {BUF}", "BUF", buf);
+        sd_journal_print(LOG_INFO, "In Ras decoder\n");
```
Please remove this debug log

akky16 replied:
Removed, done

akky16 replied:
@abinayaddhandapani @ojayanth 

Are you looking for the conditional compilation for requestGPIOEvents API or is this required for all the APIs modified in this PR.

abinayaddhandapani commented on src/apml_manager.cpp:
```diff
@@ -1506,8 +1374,9 @@ bool Manager::decodeInterrupt(uint8_t socNum)
-            else if (buf & fatalError)
+            else if (src & APML_FATAL_ALERT)
             {
+                sd_journal_print(LOG_INFO, "FATAL detected\n");
```
Please remove debug logs

akky16 commented on include/base_manager.hpp:
```diff
@@ -47,14 +47,22 @@ class Manager
+     the derived class.
+     *
+     */
+    virtual void releaseUdevReSrc() = 0;
```
ok, thank you for the info

akky16 commented on include/apml_manager.hpp:
```diff
@@ -5,6 +5,8 @@
 extern "C"
 {
 #include "esmi_mailbox.h"
+#include "linux/amd-apml.h"
```
Thanks Jaynath for the feedback, we have updated the commit message and fixed the code formatting.

ojayanth replied:
Thanks for the detailed and well-structured commit message — it’s very helpful.

Currently, the GPIO Monitor function is tracking the following GPIOs. Could you confirm whether the proposed design fully covers all the intended requirements, or if it is limited to a specific subset? Please review and clarify.

Reference: GPIO definition here https://github.com/AMDESE/amd-bmc-ras/blob/main/src/main.cpp#L41

I’ll continue with the code review once the high-level flow is finalized.

nchatrad replied:
Only the APML ALertL GPIOs will be covered by Alert_L module.
Monitoring of the PMIC and FPGA GPIOs is not modified by this patch.

abinayaddhandapani replied:
We may have a requirement to include PMIC ALERT lines monitoring in the RAS application, which will utilize the requestGPIOEvents API. The plan is to add the necessary GPIO lines in the amd_ras_gpio_config.json file.

Therefore, please ensure that these APIs are not removed.

In the meantime, to avoid invoking this code until the feature is fully integrated, please use conditional compilation to guard the relevant sections.

ojayanth commented on src/apml_manager.cpp:
```diff
@@ -492,46 +481,63 @@ void Manager::init()
     {
-        socketNames =
-            config["GPIO_ALERT_LINES"].get<std::vector<std::string>>();
+        sd_journal_print(LOG_INFO, "Invalid udev device\n");
```
File scope: Use lg2 based Logging 
Eg: lg2::error("Exception: {ERROR}", "ERROR", e.what());

abinayaddhandapani commented on src/apml_manager.cpp:
```diff
@@ -1411,343 +1446,272 @@ bool Manager::decodeInterrupt(uint8_t socNum)
-            if (socNum == socket0)
+        if (cpuCount == 2)
+        {
+            if ((p0AlertProcessed == true) || (p1AlertProcessed == true))
```
Ras Status register will be set for both the sockets even if error is injected in one. 

root@chalupa-4a1a:~# apml_tool 0 --readregister sbrmi 0x4c

================================= APML System Management Interface ====================================
---------------------------------
| Register      | Value         |
---------------------------------
| 0x4c          | 0x1           |
---------------------------------

========================================== End of APML SMI ============================================
root@chalupa-4a1a:~# apml_tool 1 --readregister sbrmi 0x4c

================================= APML System Management Interface ====================================
---------------------------------
| Register      | Value         |
---------------------------------
| 0x4c          | 0x1           |
---------------------------------


So please dont change the existing flow where && is used

akky16 replied:
is this expected behavior? also is this for all RAS errors, or only in case of fatal?

if this is expected for fatal, why this condition required anyway?

abinayaddhandapani replied:
@akky16  This will not only affect the RecoveryPath , but also cper files will be generated without harvesting the debug log ID for both the sockets. Please add that as well for your validation.

sathyapk committed:
Integrate APML Alert_L APIs to monitor RAS alerts

- Modify configure() to call APML API:
        The configure() function is modified to replace userspace
gpiolib with the APML API "apml_register_udev_monitor()" which
registers for uevents from the Alert_L driver associated with RAS events.

- Create alertSrcHandler() to monitor RAS alerts:
        The alertSrcHandler() function calls the APML API
"monitor_ras_alert()" to monitor RAS alert from the Alert_L driver.
A single instance of alertSrcHandler() can handle alerts from
multiple sockets. Monitoring the Alert_L GPIO and clearing
the RMI Status(0x2) and RMI RASstatus(0x4c) registers on RAS alert
is managed by the APML driver.

These change do not alter the overall structure and CPER creation.
Requesting, Binding and monitoring of GPIOs can be removed as this
functionality is now handled by the Alert_L driver. Only one thread
is required to monitor RAS alerts, regardless of the number of
sockets.

This implementation allows multiple user applications to register
for uevents and get the alert source information simultaneously.

Tested fields - Verified on Morocco system (P0 and P1).
CPER files are generated upon FATAL error injection for the socket.

Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
Signed-off-by: Sathya Priya Kumar <SathyaPriya.K@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>

sathyapk committed:
Add support to handle MCA shutdown interrupts

Support receiving interrupt for MCA Shutdown events

On a MCA shutdown error , ALERT_L will be triggered.
Alert_L module checks the RasStatusReg[Bit6] to identify
shutdown event and sends the event to user space.

Upon receiving the event, Ras app follows ths fatal error path to
harvest MCA banks and create CPER record and take recovery action.

Tested fields: Unit test done for shutdown error in Congo.

Signed-off-by: Sathya Priya Kumar <SathyaPriya.K@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>
Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>

raja-gade commented:
Closing this PR as it no longer needed
